File: packages/create-x402-service/templates/go-nethttp/internal/x402/registry_client.go
package x402

import (
	"bytes"
	"encoding/json"
	"fmt"
	"log"
	"net/http"
)

type ServiceRegistration struct {
	Name            string   `json:"name"`
	Description     string   `json:"description"`
	URL             string   `json:"url"`
	Category        string   `json:"category"`
	Pricing         Pricing  `json:"pricing"`
	WalletAddress   string   `json:"walletAddress"`
	Network         string   `json:"network"`
	AcceptedTokens  []string `json:"acceptedTokens"`
	Capabilities    []string `json:"capabilities"`
	Tags            []string `json:"tags"`
}

type Pricing struct {
	Amount   float64 `json:"amount"`
	Currency string  `json:"currency"`
}

type RegistrationResponse struct {
	ServiceID string `json:"serviceId"`
}

type RegistryClient struct {
	registryURL string
	client      *http.Client
	serviceID   string
}

func NewRegistryClient(registryURL string) *RegistryClient {
	return &RegistryClient{
		registryURL: registryURL,
		client:      &http.Client{},
	}
}

func (rc *RegistryClient) RegisterService(registration ServiceRegistration) (string, error) {
	body, err := json.Marshal(registration)
	if err != nil {
		return "", err
	}

	resp, err := rc.client.Post(
		fmt.Sprintf("%s/services/register", rc.registryURL),
		"application/json",
		bytes.NewBuffer(body),
	)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	var result RegistrationResponse
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return "", err
	}

	rc.serviceID = result.ServiceID
	log.Printf("Service registered with x402 registry: %s", rc.serviceID)

	return rc.serviceID, nil
}

func (rc *RegistryClient) SetServiceStatus(status string) error {
	if rc.serviceID == "" {
		return fmt.Errorf("service not registered")
	}

	body, err := json.Marshal(map[string]string{"status": status})
	if err != nil {
		return err
	}

	req, err := http.NewRequest(
		"PATCH",
		fmt.Sprintf("%s/services/%s/status", rc.registryURL, rc.serviceID),
		bytes.NewBuffer(body),
	)
	if err != nil {
		return err
	}

	req.Header.Set("Content-Type", "application/json")

	resp, err := rc.client.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	log.Printf("Service status updated: %s -> %s", rc.serviceID, status)
	return nil
}

func (rc *RegistryClient) Heartbeat() error {
	if rc.serviceID == "" {
		return fmt.Errorf("service not registered")
	}

	resp, err := rc.client.Post(
		fmt.Sprintf("%s/services/%s/heartbeat", rc.registryURL, rc.serviceID),
		"application/json",
		nil,
	)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	log.Printf("Heartbeat sent to registry: %s", rc.serviceID)
	return nil
}

func (rc *RegistryClient) GetServiceID() string {
	return rc.serviceID
}

File: packages/create-x402-service/templates/rust-actix/src/x402/registry_client.rs
use reqwest::Client;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ServiceRegistration {
    pub name: String,
    pub description: String,
    pub url: String,
    pub category: String,
    pub pricing: Pricing,
    pub wallet_address: String,
    pub network: String,
    pub accepted_tokens: Vec<String>,
    pub capabilities: Vec<String>,
    pub tags: Vec<String>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct Pricing {
    pub amount: f64,
    pub currency: String,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RegistrationResponse {
    pub service_id: String,
}

pub struct RegistryClient {
    registry_url: String,
    client: Client,
    service_id: Option<String>,
}

impl RegistryClient {
    pub fn new(registry_url: String) -> Self {
        Self {
            registry_url,
            client: Client::new(),
            service_id: None,
        }
    }

    pub async fn register_service(&mut self, registration: ServiceRegistration) -> Result<String, Box<dyn std::error::Error>> {
        let response = self
            .client
            .post(format!("{}/services/register", self.registry_url))
            .json(&registration)
            .send()
            .await?
            .json::<RegistrationResponse>()
            .await?;

        self.service_id = Some(response.service_id.clone());
        log::info!("Service registered with x402 registry: {}", response.service_id);

        Ok(response.service_id)
    }

    pub async fn set_service_status(&self, status: &str) -> Result<(), Box<dyn std::error::Error>> {
        let service_id = self.service_id.as_ref().ok_or("Service not registered")?;

        let mut body = HashMap::new();
        body.insert("status", status);

        self.client
            .patch(format!("{}/services/{}/status", self.registry_url, service_id))
            .json(&body)
            .send()
            .await?;

        log::info!("Service status updated: {} -> {}", service_id, status);
        Ok(())
    }

    pub async fn heartbeat(&self) -> Result<(), Box<dyn std::error::Error>> {
        let service_id = self.service_id.as_ref().ok_or("Service not registered")?;

        self.client
            .post(format!("{}/services/{}/heartbeat", self.registry_url, service_id))
            .send()
            .await?;

        log::debug!("Heartbeat sent to registry: {}", service_id);
        Ok(())
    }

    pub fn get_service_id(&self) -> Option<&String> {
        self.service_id.as_ref()
    }
}
